dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,385 @@
+import math
+import numpy as np
+import scipy.stats as stats
+from functools import partial
+from pprint import pprint
+from toolz import reduce
+import toolz
+import warnings
+from timeit import default_timer
+
+import sklearn
+from dask_ml.model_selection._split import train_test_split
+
+from sklearn import clone
+from sklearn.model_selection import ParameterSampler
+
+import dask.array as da
+from .model_selection import DaskBaseSearchCV, GridSearchCV, _RETURN_TRAIN_SCORE_DEFAULT
+
+
+def _train(model, X, y, X_val, y_val, max_iter=1, dry_run=False, classes=None,
+           s=1, i=1, k='_', verbose=False, **fit_kwargs):
+    """
+    Train function. Returns validation loss.
+
+    Takes max_iters as parameters. THese are treated as a "scarce resource" --
+    Hyperband is trying to minimize the total iters use for max_iters.
+    """
+    start_time = default_timer()
+    for iter in range(int(max_iter)):
+        if verbose:
+            msg = ("Training model {k} in bracket s={s} iteration {i}. "
+                   "This model is {percent:.1f}% trained for this bracket")
+            print(msg.format(iter=iter, k=k, s=s, i=i, percent=iter * 100.0 / max_iter))
+        if not dry_run:
```

Comment:
Note to self: what is dry_run for? Seems to be just for testing. Let's see if we can refactor the implementation / tests to make dry_run unnecessary.

Reply:
`dry_run` is for `Hyperband.info`, which runs Hyperband without calling `model.fit`. `info` is a useful metric to see how many calls to `partial_fit` will be made.